util/types/datum.go

```diff
@@ -1048,20 +1048,32 @@ func (d *Datum) convertToMysqlDecimal(sc *variable.StatementContext, target *Fie
 	default:
 		return invalidConv(d, target.Tp)
 	}
-	if target.Flen != UnspecifiedLength && target.Decimal != UnspecifiedLength {
+	dec, err = produceDecWithSpecifiedTp(dec, target, sc)
+	ret.SetValue(dec)
+	return ret, errors.Trace(err)
+}
+
+// produceDecToSpecifiedTp produces a new decimal according to `Tp`.
+func produceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *variable.StatementContext) (_ *MyDecimal, err error) {
+	flen, decimal := tp.Flen, tp.Decimal
+	val, err := dec.ToFloat64()
```

Comment:
Convert decimal to float64 is expensive, we only need to know if the decimal is zero.
Adding an `IsZero` method for `MyDecimal` is better.